src/Core/Grid/Data/Factory/DoctrineGridDataFactory.php

```diff
@@ -93,9 +95,39 @@ public function getData(SearchCriteriaInterface $searchCriteria)
             'search_criteria' => $searchCriteria,
         ]);
 
-        $records = $searchQueryBuilder->execute()->fetchAll();
         $recordsTotal = (int) $countQueryBuilder->execute()->fetch(PDO::FETCH_COLUMN);
+        $searchOffset = (int) $searchCriteria->getOffset();
+
+        // If offset is out of range, we compute and apply a valid offset.
+        if (Pagination::isOffsetOutOfRange($recordsTotal, $searchOffset)) {
+            // @TODO : those cases should be simplified with some refactoring cf. https://github.com/PrestaShop/PrestaShop/issues/29131
+
+            $searchLimit = $searchCriteria->getLimit();
+            $inRangeOffset = Pagination::computeValidOffset($recordsTotal, $searchOffset, $searchLimit);
+
+            // @TODO : add INFO log to prevent that we overwrite offset from `$searchCriteria->getOffset()` to `$inRangeOffset`
 
+            if ($searchCriteria instanceof Filters) {
+                $newSearchCriteria = clone $searchCriteria;
+                $newSearchCriteria->add(['offset' => $inRangeOffset]);
+            } else {
+                // @TODO : add INFO/WARN log to prevent that we replaced `get_class($searchCriteria)` by `Filters` search criteria
+
+                $newSearchCriteria = new Filters(
```

Comment:
Since the initial criteria was not a filter I'm not sure we should use `Filters` as the replacement but maybe instead a more generic class like `SearchCriteria`